producer/writer/message_writer.go

```diff
@@ -0,0 +1,378 @@
+			Counter("write"),
+		writeAfterCutoff: scope.
+			Tagged(map[string]string{"reason": "after-cutoff"}).
+			Counter("invalid-write"),
+		writeBeforeCutover: scope.
+			Tagged(map[string]string{"reason": "before-cutover"}).
+			Counter("invalid-write"),
+	}
+}
+
+type messageWriterImpl struct {
+	sync.RWMutex
+
+	replicatedShardID uint64
+	mPool             messagePool
+	opts              Options
+
+	msgID           uint64
+	queue           *list.List
+	consumerWriters map[string]consumerWriter
+	acks            *acks
+	cutOffNanos     int64
+	cutOverNanos    int64
+	isClosed        bool
+	doneCh          chan struct{}
+	wg              sync.WaitGroup
+	m               messageWriterMetrics
+
+	nowFn clock.NowFn
+}
+
+func newMessageWriter(
+	replicatedShardID uint64,
+	mPool messagePool,
+	opts Options,
+) messageWriter {
+	if opts == nil {
+		opts = NewOptions()
+	}
+	return &messageWriterImpl{
+		replicatedShardID: replicatedShardID,
+		mPool:             mPool,
+		opts:              opts,
+		msgID:             0,
+		queue:             list.New(),
+		consumerWriters:   make(map[string]consumerWriter),
+		acks:              newAckHelper(defaultAckMapSize),
+		cutOffNanos:       0,
+		cutOverNanos:      0,
+		isClosed:          false,
+		doneCh:            make(chan struct{}),
+		m:                 newMessageWriterMetrics(opts.InstrumentOptions().MetricsScope()),
+		nowFn:             time.Now,
+	}
+}
+
+func (w *messageWriterImpl) Write(rd producer.RefCountedData) {
+	nowNanos := w.nowFn().UnixNano()
+	w.RLock()
+	isValid := w.isValidWriteWithLock(nowNanos)
+	w.RUnlock()
+	if !isValid {
+		return
+	}
+	rd.IncRef()
+	msg := w.mPool.Get()
+
+	w.Lock()
+	w.msgID++
+	meta := metadata{
+		shard: w.replicatedShardID,
+		id:    w.msgID,
+	}
+	msg.Reset(meta, rd)
+	w.acks.add(meta, msg)
+	w.writeWithLock(msg, nowNanos)
+	w.queue.PushBack(msg)
+	w.Unlock()
+
+	w.m.writeNew.Inc(1)
+}
+
+func (w *messageWriterImpl) isValidWriteWithLock(nowNanos int64) bool {
+	if w.cutOffNanos > 0 && nowNanos >= w.cutOffNanos {
+		w.m.writeAfterCutoff.Inc(1)
+		return false
+	}
+	if w.cutOverNanos > 0 && nowNanos < w.cutOverNanos {
+		w.m.writeBeforeCutover.Inc(1)
+		return false
+	}
+	return true
+}
+
+func (w *messageWriterImpl) writeWithLock(m *message, nowNanos int64) {
+	m.IncReads()
+	msg, isValid := m.Marshaler()
+	if !isValid {
+		m.DecReads()
+		return
+	}
+	written := false
+	for _, cw := range w.consumerWriters {
+		if err := cw.Write(msg); err != nil {
+			w.m.consumerWriteError.Inc(1)
+			continue
+		}
+		written = true
+		w.m.writeSuccess.Inc(1)
+		break
+	}
+	m.DecReads()
+
+	if !written {
+		// Could not be written to any consumer, will retry later.
+		w.m.writeError.Inc(1)
+	}
+	m.SetRetryNanos(w.nextRetryNanos(m, nowNanos))
+}
+
+// TODO: make retry time strategy configurable.
+func (w *messageWriterImpl) nextRetryNanos(m *message, nowNanos int64) int64 {
+	return nowNanos + int64(m.RetriedTimes())*int64(w.opts.MessageRetryBackoff())
```

Comment:
Do you need any jittering here? Otherwise if you have a batch of messages that are failed due to a network blip, they'll be retried in batch next time as well.

Reply:
yeah could add some jitter